test/Interpreter/error_self_conformance.swift

```diff
@@ -0,0 +1,44 @@
+// RUN: %target-jit-run %s 2>&1 | %FileCheck %s
+// REQUIRES: swift_interpreter
+
+struct BoxedError<T: Error> {
+  var contents: T
+}
+
+func takeBoxedError<T>(error: BoxedError<T>) -> Error {
+  return error.contents
+}
+
+func makeBoxedError<T: Error>(error: T) -> BoxedError<Error> {
+  return BoxedError(contents: error)
+}
+
+enum MyError: Error {
+  case nothingImportant
+  case reallyImportant(String)
+}
+
+func castValueToError<T>(error: T) -> Error? {
+  return error as? Error
+}
```

Comment:
You'll need to update `swift::_conformsToProtocol` in the runtime, because it has a path that will need to know that an error existential conforms to the Error protocol. (Right now it only considers `@objc` existentials to conform to their protocols). You can exercise the code by evaluating a conditional conformance at runtime.

Reply:
It actually Just Works, probably because I'm generating an ordinary conformance with a conformance descriptor whose conforming type is linked correctly back to the protocol (by a cavalcade of different PRs).  But as discussed privately, I'll flesh out the tests for various cases of this.